File: backend/internal/domain/validation.go
package domain

import (
	"regexp"
)

func CleanCPFCNPJ(s string) string {
	reg := regexp.MustCompile(`[^0-9]`)
	return reg.ReplaceAllString(s, "")
}

func IsValidCPFCNPJ(s string) bool {
	s = CleanCPFCNPJ(s)
	if len(s) == 11 {
		return IsValidCPF(s)
	} else if len(s) == 14 {
		return IsValidCNPJ(s)
	}
	return false
}

func IsValidCPF(cpf string) bool {
	if len(cpf) != 11 {
		return false
	}

	allSame := true
	for i := 1; i < 11; i++ {
		if cpf[i] != cpf[0] {
			allSame = false
			break
		}
	}
	if allSame {
		return false
	}

	sum := 0
	for i := 0; i < 9; i++ {
		sum += int(cpf[i]-'0') * (10 - i)
	}
	rem := sum % 11
	d1 := 0
	if rem >= 2 {
		d1 = 11 - rem
	}
	if int(cpf[9]-'0') != d1 {
		return false
	}

	sum = 0
	for i := 0; i < 10; i++ {
		sum += int(cpf[i]-'0') * (11 - i)
	}
	rem = sum % 11
	d2 := 0
	if rem >= 2 {
		d2 = 11 - rem
	}
	return int(cpf[10]-'0') == d2
}

func IsValidCNPJ(cnpj string) bool {
	if len(cnpj) != 14 {
		return false
	}

	allSame := true
	for i := 1; i < 14; i++ {
		if cnpj[i] != cnpj[0] {
			allSame = false
			break
		}
	}
	if allSame {
		return false
	}

	w1 := []int{5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2}
	w2 := []int{6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2}

	sum := 0
	for i := 0; i < 12; i++ {
		sum += int(cnpj[i]-'0') * w1[i]
	}
	rem := sum % 11
	d1 := 0
	if rem >= 2 {
		d1 = 11 - rem
	}
	if int(cnpj[12]-'0') != d1 {
		return false
	}

	sum = 0
	for i := 0; i < 13; i++ {
		sum += int(cnpj[i]-'0') * w2[i]
	}
	rem = sum % 11
	d2 := 0
	if rem >= 2 {
		d2 = 11 - rem
	}
	return int(cnpj[13]-'0') == d2
}

File: backend/Models/Validation.cs
using System.Text.RegularExpressions;

namespace PortalProposta.Models;

public static class Validation
{
    public static string CleanCPFCNPJ(string? s) =>
        s == null ? string.Empty : Regex.Replace(s, @"[^0-9]", "");

    public static bool IsValidCPFCNPJ(string s)
    {
        s = CleanCPFCNPJ(s);
        return s.Length switch
        {
            11 => IsValidCPF(s),
            14 => IsValidCNPJ(s),
            _ => false
        };
    }

    public static bool IsValidCPF(string cpf)
    {
        if (cpf.Length != 11) return false;
        if (cpf.Distinct().Count() == 1) return false;

        var sum = 0;
        for (var i = 0; i < 9; i++) sum += (cpf[i] - '0') * (10 - i);
        var d1 = sum % 11 >= 2 ? 11 - sum % 11 : 0;
        if ((cpf[9] - '0') != d1) return false;

        sum = 0;
        for (var i = 0; i < 10; i++) sum += (cpf[i] - '0') * (11 - i);
        var d2 = sum % 11 >= 2 ? 11 - sum % 11 : 0;
        return (cpf[10] - '0') == d2;
    }

    public static bool IsValidCNPJ(string cnpj)
    {
        if (cnpj.Length != 14) return false;
        if (cnpj.Distinct().Count() == 1) return false;

        var w1 = new[] { 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };
        var w2 = new[] { 6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };

        var sum = 0;
        for (var i = 0; i < 12; i++) sum += (cnpj[i] - '0') * w1[i];
        var d1 = sum % 11 >= 2 ? 11 - sum % 11 : 0;
        if ((cnpj[12] - '0') != d1) return false;

        sum = 0;
        for (var i = 0; i < 13; i++) sum += (cnpj[i] - '0') * w2[i];
        var d2 = sum % 11 >= 2 ? 11 - sum % 11 : 0;
        return (cnpj[13] - '0') == d2;
    }
}
